--- a/src/systemd/meson.build	2018-03-03 22:56:45.111894767 +0100
+++ b/src/systemd/meson.build	2018-03-03 22:58:09.791897314 +0100
@@ -22,9 +22,6 @@
         sd-daemon.h
         sd-event.h
         sd-id128.h
-        sd-journal.h
-        sd-login.h
-        sd-messages.h
 '''.split()
 
 # https://github.com/mesonbuild/meson/issues/1633
--- a/src/libsystemd/meson.build	2018-04-13 12:53:10.169234014 +0200
+++ b/src/libsystemd/meson.build	2018-04-13 12:53:39.005234881 +0200
@@ -65,7 +65,7 @@
         sd-path/sd-path.c
         sd-resolve/sd-resolve.c
         sd-utf8/sd-utf8.c
-'''.split()) + id128_sources + sd_daemon_c + sd_event_c + sd_login_c
+'''.split()) + id128_sources + sd_daemon_c + sd_event_c
 
 libsystemd_static = static_library(
         'systemd',
